expression/integration_test.go

```diff
@@ -1280,8 +1280,15 @@ func (s *testIntegrationSuite) TestTimeBuiltin(c *C) {
 	// for quarter
 	result = tk.MustQuery(`select quarter("2012-00-20"), quarter("2012-01-21"), quarter("2012-03-22"), quarter("2012-05-23"), quarter("2012-08-24"), quarter("2012-09-25"), quarter("2012-11-26"), quarter("2012-12-27");`)
 	result.Check(testkit.Rows("0 1 1 2 3 3 4 4"))
-	result = tk.MustQuery(`select quarter("2012-14-20"), quarter("0000-00-00"), quarter("aa"), quarter(null), quarter(11), quarter(12.99);`)
-	result.Check(testkit.Rows("<nil> <nil> <nil> <nil> <nil> <nil>"))
+	result = tk.MustQuery(`select quarter("2012-14-20"), quarter("aa"), quarter(null), quarter(11), quarter(12.99);`)
+	result.Check(testkit.Rows("<nil> <nil> <nil> <nil> <nil>"))
+	result = tk.MustQuery(`select quarter("0000-00-00"), quarter("0000-00-00 00:00:00");`)
+	result.Check(testkit.Rows("<nil> <nil>"))
+	tk.MustQuery("show warnings").Check(testutil.RowsWithSep("|",
+		"Warning|1292|Incorrect datetime value: '0000-00-00 00:00:00.000000'",
+		"Warning|1292|Incorrect datetime value: '0000-00-00 00:00:00.000000'"))
+	result = tk.MustQuery(`select quarter(0), quarter(0.0);`)
+	result.Check(testkit.Rows("0 0"))
```

Comment:
The issue also describes the `DATE_FORMAT` problem, could you fix it together?

Reply:
Could you provide any examples for `DATE_FORMAT`?

Original issue says that `QUARTER` function returns `NULL` in tidb for 0/0.0 while mysql returns `0`.

I tried the same values for `DATE_FORMAT` and it looks like that the issue is the opposite — tidb returns `0` while mysql returns `NULL`:

**Schema (MySQL v5.7)**
```
mysql> select DATE_FORMAT(0, '%W'), DATE_FORMAT(0.0, '%W'), DATE_FORMAT('0', '%W'), DATE_FORMAT('0.0', '%W');

| DATE_FORMAT(0, '%W') | DATE_FORMAT(0.0, '%W') | DATE_FORMAT('0', '%W') | DATE_FORMAT('0.0', '%W') |
| -------------------- | ---------------------- | ---------------------- | ------------------------ |
|                 null |                   null |                   null |                     null |

mysql> show warnings;

| Level   | Code | Message                         |
| ------- | ---- | ------------------------------- |
| Warning | 1292 | Incorrect datetime value: '0'   |
| Warning | 1292 | Incorrect datetime value: '0.0' |
```

[View on DB Fiddle](https://www.db-fiddle.com/f/x4Cja21Jtv21H8qWoMdm9r/2)

**TiDB**:

```
mysql> select DATE_FORMAT(0, '%W'), DATE_FORMAT(0.0, '%W'), DATE_FORMAT('0', '%W'), DATE_FORMAT('0.0', '%W');
+----------------------+------------------------+------------------------+--------------------------+
| DATE_FORMAT(0, '%W') | DATE_FORMAT(0.0, '%W') | DATE_FORMAT('0', '%W') | DATE_FORMAT('0.0', '%W') |
+----------------------+------------------------+------------------------+--------------------------+
| 0                    | 0                      | NULL                   | NULL                     |
+----------------------+------------------------+------------------------+--------------------------+
1 row in set, 2 warnings (0.00 sec)

mysql> 
mysql> show warnings;
+---------+------+----------------------------+
| Level   | Code | Message                    |
+---------+------+----------------------------+
| Warning | 1292 | invalid time format: '0'   |
| Warning | 1292 | invalid time format: '0.0' |
+---------+------+----------------------------+
2 rows in set (0.00 sec)
```

Reply:
An example is `select DATE_FORMAT(0, 0);`

in TiDB, it returns NULL with a warning:
```mysql
mysql> select DATE_FORMAT(0, 0);
+-------------------+
| DATE_FORMAT(0, 0) |
+-------------------+
| NULL              |
+-------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+--------------------------------------------------------+
| Level   | Code | Message                                                |
+---------+------+--------------------------------------------------------+
| Warning | 1292 | Incorrect datetime value: '0000-00-00 00:00:00.000000' |
+---------+------+--------------------------------------------------------+
1 row in set (0.00 sec)
```

in MySQL, it returns 0 with no warning:
```mysql
mysql> select DATE_FORMAT(0, 0);
+-------------------+
| DATE_FORMAT(0, 0) |
+-------------------+
| 0                 |
+-------------------+
1 row in set (0.00 sec)
```

Reply:
Got it. Thanks.
Added a fix + test.